File: testers/sqlserver/scala/src/scala/testdb/AllTypesTest.scala
package testdb

import org.junit.Assert._
import org.junit.Test
import testdb.all_scalar_types._
import testdb.customtypes.Defaulted

import java.time._
import java.util.UUID

/** Comprehensive tests for all SQL Server scalar data types. Tests the all_scalar_types table which has 38 columns covering all supported SQL Server types.
  */
class AllTypesTest {
  private val repo = AllScalarTypesRepoImpl()

  def createSampleRow(): AllScalarTypesRowUnsaved = {
    AllScalarTypesRowUnsaved(
      colTinyint = Some(dev.typr.foundations.data.Uint1(255)),
      colSmallint = Some(1000.toShort),
      colInt = Some(100000),
      colBigint = Some(10000000000L),
      colDecimal = Some(BigDecimal("12345.6789")),
      colNumeric = Some(BigDecimal("999.99")),
      colMoney = Some(BigDecimal("922337203685477.5807")),
      colSmallmoney = Some(BigDecimal("214748.3647")),
      colReal = Some(3.14f),
      colFloat = Some(2.718281828),
      colBit = Some(true),
      colChar = Some("char_val  "),
      colVarchar = Some("varchar_value"),
      colVarcharMax = Some("varchar max content"),
      colText = Some("text content"),
      colNchar = Some("nchar_val "),
      colNvarchar = Some("nvarchar_value 中文"),
      colNvarcharMax = Some("nvarchar max content 日本語"),
      colNtext = Some("ntext content"),
      colBinary = Some(Array[Byte](1, 2, 3, 4, 5, 6, 7, 8, 9, 10)),
      colVarbinary = Some(Array[Byte](1, 2, 3)),
      colVarbinaryMax = Some(Array[Byte](4, 5, 6, 7, 8)),
      colImage = Some(Array[Byte](9, 10, 11)),
      colDate = Some(LocalDate.of(2025, 1, 15)),
      colTime = Some(LocalTime.of(14, 30, 45, 123456700)),
      colDatetime = Some(LocalDateTime.of(2025, 1, 15, 14, 30, 45)),
      colSmalldatetime = Some(LocalDateTime.of(2025, 1, 15, 14, 30)),
      colDatetime2 = Some(LocalDateTime.of(2025, 1, 15, 14, 30, 45, 123456700)),
      colDatetimeoffset = Some(OffsetDateTime.of(2025, 1, 15, 14, 30, 45, 0, ZoneOffset.ofHours(-5))),
      colUniqueidentifier = Some(UUID.fromString("550e8400-e29b-41d4-a716-446655440000")),
      colXml = Some(dev.typr.foundations.data.Xml("<root><element>value</element></root>")),
      colJson = Some(dev.typr.foundations.data.Json("{\"key\": \"value\"}")),
      colHierarchyid = None,
      colGeography = None,
      colGeometry = None,
      colNotNull = Defaulted.Provided("test_value")
    )
  }

  @Test
  def testInsertAndSelectAllTypes(): Unit = withConnection {
    val row = createSampleRow()
    val inserted = repo.insert(row)

    assertNotNull(inserted)
    assertNotNull(inserted.id)
    assertEquals(row.colTinyint, inserted.colTinyint)
    assertEquals(row.colSmallint, inserted.colSmallint)
    assertEquals(row.colInt, inserted.colInt)
    assertEquals(row.colBigint, inserted.colBigint)
    assertEquals(row.colBit, inserted.colBit)
    assertEquals(row.colVarchar, inserted.colVarchar)
    assertEquals(row.colNvarchar, inserted.colNvarchar)
    assertEquals(row.colDate, inserted.colDate)
    assertEquals(row.colUniqueidentifier, inserted.colUniqueidentifier)

    assertNotNull(inserted.colRowversion)

    val found = repo.selectById(inserted.id)
    assertTrue(found.isDefined)
    assertEquals(inserted.id, found.get.id)
  }

  @Test
  def testUpdateAllTypes(): Unit = withConnection {
    val row = createSampleRow()
    val inserted = repo.insert(row)

    val updated = inserted.copy(
      colVarchar = Some("updated_varchar"),
      colNvarchar = Some("updated_nvarchar"),
      colDecimal = Some(BigDecimal("9999.9999")),
      colBit = Some(false)
    )

    val wasUpdated = repo.update(updated)
    assertTrue(wasUpdated)

    val found = repo.selectById(inserted.id).get
    assertEquals(Some("updated_varchar"), found.colVarchar)
    assertEquals(Some("updated_nvarchar"), found.colNvarchar)
    assertEquals(Some(BigDecimal("9999.9999")), found.colDecimal)
    assertEquals(Some(false), found.colBit)
  }

  @Test
  def testDeleteAllTypes(): Unit = withConnection {
    val row = createSampleRow()
    val inserted = repo.insert(row)

    val deleted = repo.deleteById(inserted.id)
    assertTrue(deleted)

    val found = repo.selectById(inserted.id)
    assertFalse(found.isDefined)
  }

  @Test
  def testInsertWithNulls(): Unit = withConnection {
    val row = AllScalarTypesRowUnsaved()
    val inserted = repo.insert(row)

    assertNotNull(inserted)
    assertNotNull(inserted.id)

    assertTrue(inserted.colTinyint.isEmpty)
    assertTrue(inserted.colSmallint.isEmpty)
    assertTrue(inserted.colInt.isEmpty)
    assertTrue(inserted.colBit.isEmpty)
    assertTrue(inserted.colVarchar.isEmpty)
    assertTrue(inserted.colDate.isEmpty)
    assertTrue(inserted.colUniqueidentifier.isEmpty)

    assertEquals("default_value", inserted.colNotNull)
    assertNotNull(inserted.colRowversion)
  }

  @Test
  def testIntegerTypes(): Unit = withConnection {
    val row = createSampleRow()
    val inserted = repo.insert(row)

    assertEquals(Some(dev.typr.foundations.data.Uint1(255)), inserted.colTinyint)
    assertEquals(Some(1000.toShort), inserted.colSmallint)
    assertEquals(Some(100000), inserted.colInt)
    assertEquals(Some(10000000000L), inserted.colBigint)
  }

  @Test
  def testMoneyTypes(): Unit = withConnection {
    val row = createSampleRow()
    val inserted = repo.insert(row)

    assertEquals(Some(BigDecimal("922337203685477.5807")), inserted.colMoney)
    assertEquals(Some(BigDecimal("214748.3647")), inserted.colSmallmoney)
  }

  @Test
  def testDateTimeTypes(): Unit = withConnection {
    val row = createSampleRow()
    val inserted = repo.insert(row)

    assertEquals(Some(LocalDate.of(2025, 1, 15)), inserted.colDate)
    assertNotNull(inserted.colTime)
    assertNotNull(inserted.colDatetime)
    assertNotNull(inserted.colDatetime2)
    assertNotNull(inserted.colDatetimeoffset)
  }

  @Test
  def testUuidType(): Unit = withConnection {
    val uuid = UUID.randomUUID()
    val row = createSampleRow().copy(colUniqueidentifier = Some(uuid))
    val inserted = repo.insert(row)

    assertEquals(Some(uuid), inserted.colUniqueidentifier)
  }

  @Test
  def testXmlType(): Unit = withConnection {
    val xml = dev.typr.foundations.data.Xml("<person><name>John</name><age>30</age></person>")
    val row = createSampleRow().copy(colXml = Some(xml))
    val inserted = repo.insert(row)

    assertTrue(inserted.colXml.isDefined)
    assertTrue(inserted.colXml.get.value.contains("John"))
  }

  @Test
  def testJsonType(): Unit = withConnection {
    val json = dev.typr.foundations.data.Json("{\"name\": \"test\", \"values\": [1, 2, 3]}")
    val row = createSampleRow().copy(colJson = Some(json))
    val inserted = repo.insert(row)

    assertTrue(inserted.colJson.isDefined)
    assertTrue(inserted.colJson.get.value.contains("name"))
  }

  @Test
  def testBinaryTypes(): Unit = withConnection {
    val binaryData = Array[Byte](0x00, 0x01, 0x02, 0xff.toByte, 0xfe.toByte)
    val row = createSampleRow().copy(colVarbinary = Some(binaryData))
    val inserted = repo.insert(row)

    assertTrue(inserted.colVarbinary.isDefined)
    assertArrayEquals(binaryData, inserted.colVarbinary.get)
  }

  @Test
  def testUnicodeStrings(): Unit = withConnection {
    val unicode = "Hello 世界 🌍 مرحبا"
    val row = createSampleRow().copy(colNvarchar = Some(unicode))
    val inserted = repo.insert(row)

    assertEquals(Some(unicode), inserted.colNvarchar)
  }

  @Test
  def testRowversionAutoGenerated(): Unit = withConnection {
    val row1 = repo.insert(createSampleRow())
    val row2 = repo.insert(createSampleRow())

    assertNotNull(row1.colRowversion)
    assertNotNull(row2.colRowversion)
    assertFalse(java.util.Arrays.equals(row1.colRowversion, row2.colRowversion))
  }
}

File: testers/sqlite/kotlin/src/kotlin/testdb/AllTypesTest.kt
package testdb

import dev.typr.foundations.data.Json
import org.junit.Assert.*
import org.junit.Test
import testdb.all_scalar_types.*
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.util.UUID

class AllTypesTest {
    private val repo = AllScalarTypesRepoImpl()

    private fun createSampleRow(id: Long): AllScalarTypesRow = AllScalarTypesRow(
        AllScalarTypesId(id),
        42.toByte(),
        1000.toShort(),
        100000,
        10000000000L,
        9223372036854775000L,
        true,
        3.14159,
        2.718281828,
        1.5f,
        BigDecimal("12345.67"),
        BigDecimal("999.999"),
        "text content",
        "varchar_value",
        "char5",
        "clob content",
        byteArrayOf(1, 2, 3, 4, 5),
        byteArrayOf(0x0A, 0x0B, 0x0C),
        LocalDate.of(2025, 1, 15),
        LocalTime.of(14, 30, 45),
        LocalDateTime.of(2025, 1, 15, 14, 30, 45),
        LocalDateTime.of(2025, 6, 1, 9, 0, 0),
        UUID.fromString("550e8400-e29b-41d4-a716-446655440000"),
        Json("{\"key\": \"value\"}"),
        "required_value"
    )

    @Test
    fun testInsertAndSelectAllTypes() {
        SqliteTestHelper.run { c ->
            val row = createSampleRow(1001L)
            val inserted = repo.insert(row, c)
            assertEquals(row.id, inserted.id)
            assertEquals(row.colTinyint, inserted.colTinyint)
            assertEquals(row.colBoolean, inserted.colBoolean)
            assertEquals(row.colVarchar, inserted.colVarchar)
            assertEquals(row.colDate, inserted.colDate)
            assertEquals(row.colUuid, inserted.colUuid)
            assertEquals(row.colNotNull, inserted.colNotNull)
            assertNotNull(repo.selectById(inserted.id, c))
        }
    }

    @Test
    fun testUpdateAllTypes() {
        SqliteTestHelper.run { c ->
            val row = createSampleRow(1002L)
            val inserted = repo.insert(row, c)
            val updated = inserted.copy(
                colVarchar = "updated_varchar",
                colDecimal = BigDecimal("999.99"),
                colBoolean = false
            )
            assertTrue(repo.update(updated, c))
            val found = repo.selectById(inserted.id, c)!!
            assertEquals("updated_varchar", found.colVarchar)
            assertEquals(false, found.colBoolean)
        }
    }

    @Test
    fun testDeleteAllTypes() {
        SqliteTestHelper.run { c ->
            val inserted = repo.insert(createSampleRow(1003L), c)
            assertTrue(repo.deleteById(inserted.id, c))
            assertNull(repo.selectById(inserted.id, c))
        }
    }

    @Test
    fun testInsertWithNulls() {
        SqliteTestHelper.run { c ->
            val row = AllScalarTypesRow(
                AllScalarTypesId(1004L),
                null, null, null, null, null, null, null, null, null, null, null, null, null, null, null,
                null, null, null, null, null, null, null, null,
                "required_only"
            )
            val inserted = repo.insert(row, c)
            assertEquals("required_only", inserted.colNotNull)
            assertNull(inserted.colTinyint)
            assertNull(inserted.colUuid)
            assertNull(inserted.colJson)
        }
    }

    @Test
    fun testIntegerTypes() {
        SqliteTestHelper.run { c ->
            val inserted = repo.insert(createSampleRow(1010L), c)
            assertEquals(42.toByte(), inserted.colTinyint)
            assertEquals(1000.toShort(), inserted.colSmallint)
            assertEquals(100000, inserted.colInt)
            assertEquals(10000000000L, inserted.colInteger)
            assertEquals(9223372036854775000L, inserted.colBigint)
        }
    }

    @Test
    fun testBooleanType() {
        SqliteTestHelper.run { c ->
            val inserted = repo.insert(createSampleRow(1011L), c)
            assertEquals(true, inserted.colBoolean)
            repo.update(inserted.copy(colBoolean = false), c)
            assertEquals(false, repo.selectById(inserted.id, c)!!.colBoolean)
        }
    }

    @Test
    fun testFloatingPointTypes() {
        SqliteTestHelper.run { c ->
            val inserted = repo.insert(createSampleRow(1012L), c)
            assertEquals(3.14159, inserted.colReal!!, 0.0)
            assertEquals(2.718281828, inserted.colDouble!!, 0.0)
            assertEquals(1.5f, inserted.colFloat!!, 0.0f)
        }
    }

    @Test
    fun testDecimalAndNumeric() {
        SqliteTestHelper.run { c ->
            val inserted = repo.insert(createSampleRow(1013L), c)
            assertEquals(0, BigDecimal("12345.67").compareTo(inserted.colDecimal!!))
            assertEquals(0, BigDecimal("999.999").compareTo(inserted.colNumeric!!))
        }
    }

    @Test
    fun testTextTypes() {
        SqliteTestHelper.run { c ->
            val inserted = repo.insert(createSampleRow(1014L), c)
            assertEquals("text content", inserted.colText)
            assertEquals("varchar_value", inserted.colVarchar)
            assertEquals("char5", inserted.colChar)
            assertEquals("clob content", inserted.colClob)
        }
    }

    @Test
    fun testBlobAndBinary() {
        SqliteTestHelper.run { c ->
            val blob = byteArrayOf(0, 1, 2, -1, -2)
            val bin = byteArrayOf(-54, -2, -70, -66)
            val inserted = repo.insert(createSampleRow(1015L).copy(colBlob = blob, colBinary = bin), c)
            assertArrayEquals(blob, inserted.colBlob)
            assertArrayEquals(bin, inserted.colBinary)
        }
    }

    @Test
    fun testDateTimeTypes() {
        SqliteTestHelper.run { c ->
            val inserted = repo.insert(createSampleRow(1016L), c)
            assertEquals(LocalDate.of(2025, 1, 15), inserted.colDate)
            assertEquals(LocalTime.of(14, 30, 45), inserted.colTime)
            assertEquals(LocalDateTime.of(2025, 1, 15, 14, 30, 45), inserted.colDatetime)
            assertEquals(LocalDateTime.of(2025, 6, 1, 9, 0, 0), inserted.colTimestamp)
        }
    }

    @Test
    fun testUuid() {
        SqliteTestHelper.run { c ->
            val uuid = UUID.randomUUID()
            val inserted = repo.insert(createSampleRow(1017L).copy(colUuid = uuid), c)
            assertEquals(uuid, inserted.colUuid)
        }
    }

    @Test
    fun testSelectAllReturnsInsertedRow() {
        SqliteTestHelper.run { c ->
            val row = createSampleRow(1019L)
            repo.insert(row, c)
            val all = repo.selectAll(c)
            assertTrue(all.size >= 2)
            assertTrue(all.any { it.id.value == 1019L })
        }
    }

    @Test
    fun testJson() {
        SqliteTestHelper.run { c ->
            val json = Json("{\"name\":\"test\"}")
            val inserted = repo.insert(createSampleRow(1018L).copy(colJson = json), c)
            assertTrue(inserted.colJson!!.value().contains("name"))
        }
    }
}
